ClientRuntimes/Java/azure-client-runtime/src/main/java/com/microsoft/azure/PagedList.java

```diff
@@ -0,0 +1,277 @@
+/**
+ *
+ * Copyright (c) Microsoft Corporation. All rights reserved.
+ * Licensed under the MIT License. See License.txt in the project root for license information.
+ *
+ */
+
+package com.microsoft.azure;
+
+import java.io.IOException;
+import java.util.Collection;
+import java.util.Iterator;
+import java.util.List;
+import java.util.ListIterator;
+import java.util.NoSuchElementException;
+
+import javax.xml.bind.DataBindingException;
+import javax.xml.ws.WebServiceException;
+
+/**
+ * Defines a list response from a paging operation. The pages are
+ * lazy initialized when an instance of this class is iterated.
+ *
+ * @param <E> the element type.
+ */
+public abstract class PagedList<E> implements List<E> {
+    /** The actual items in the list. */
+    private List<E> items;
+    /** Stores the link to get the next page of items. */
+    private String nextPageLink;
+
+    /**
+     * Creates an instance of PagedList from a {@link Page} response.
+     *
+     * @param page the {@link Page} object.
+     */
+    public PagedList(Page<E> page) {
+        items = page.getItems();
+        nextPageLink = page.getNextPageLink();
+    }
+
+    /**
+     * Override this method to load the next page of items from a next page link.
+     *
+     * @param nextPageLink the link to get the next page of items.
+     * @return the {@link Page} object storing a page of items and a link to the next page.
+     * @throws CloudException thrown if an error is raised from Azure.
+     * @throws IOException thrown if there's any failure in deserialization.
+     */
+    public abstract Page<E> loadPage(String nextPageLink) throws CloudException, IOException;
+
+    /**
+     * Keep loading the next page from the next page link until all items are loaded.
+     */
+    public void loadAll() {
+        while (nextPageLink != null) {
+            try {
+                Page<E> nextPage = loadPage(nextPageLink);
+                nextPageLink = nextPage.getNextPageLink();
+                addAll(nextPage.getItems());
+            } catch (CloudException e) {
```

Comment:
May be we could have a private method "loadNextPage" in PagedList, that wraps the user's implementation of abstract method "nextPage" along with setting nextPageLink and appending result to items

I mean something like below:

```
public abstract class PagedList<E> implements List<E> {
  ....
  // user override
   public abstract Page<E> nextPage(Sting nextPageLink) throws CloudException, IOException;

   private bool hasNextPage() {
     return nextPageLink != null
   }

   private void loadNextPage() {
     Page<E> nextPage = loadPage(nextPageLink);
     nextPageLink = nextPage.getNextPageLink();
     addAll(nextPage.getItems());
   }

   public E get(int index) {
     while (index >= items.Size() && hasMorePage()) {
       try {
         loadNextPage()
       }  catch (CloudException e) {
       ...
       } catch (IOException e) {
       ...
       }
     }  
     return items.get(index)
  }
   
  public void loadAll() {
    while (hasMorePage()) {
      try { 
        loadNextPage()
      }  catch (CloudException e) {
      ...
      } catch (IOException e) {
       ....
      }
    }
  }
  ...
}